components/script/dom/xoriginlocation.rs

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::codegen::Bindings::XOriginLocationBinding;
+use dom::bindings::codegen::Bindings::XOriginLocationBinding::XOriginLocationMethods;
+use dom::bindings::error::{Error, ErrorResult, Fallible};
+use dom::bindings::js::{JS, Root};
+use dom::bindings::reflector::DomObject;
+use dom::bindings::reflector::Reflector;
+use dom::bindings::str::DOMString;
+use dom::bindings::str::USVString;
+use dom::xoriginwindow::XOriginWindow;
+
+#[dom_struct]
+pub struct XOriginLocation {
+    reflector: Reflector,
+    window: JS<XOriginWindow>,
+}
+
+impl XOriginLocation {
+    #[allow(unsafe_code)]
+    pub fn new(window: &XOriginWindow) -> Root<XOriginLocation> {
+        let cx = window.reflector().get_cx();
+        let scope = window.reflector().get_jsobject();
+        let loc = box XOriginLocation {
+            reflector: Reflector::new(),
+            window: JS::from_ref(window),
+        };
+        unsafe {
+            XOriginLocationBinding::Wrap(cx, scope, loc)
+        }
+    }
+}
+
+impl XOriginLocationMethods for XOriginLocation {
+    // https://html.spec.whatwg.org/multipage/#dom-location-href
+    fn GetHref(&self) -> Fallible<USVString> {
+        Err(Error::Security)
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-location-href
+    fn SetHref(&self, _: USVString) -> ErrorResult {
+        // TODO: setting href on a cross-origin window should succeed?
+        Err(Error::Security)
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-location-assign
+    fn Assign(&self, _: USVString) -> Fallible<()> {
+        // TODO: setting href on a cross-origin window should succeed?
+        Err(Error::Security)
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-location-replace
+    fn Replace(&self, _: USVString) -> Fallible<()> {
+        // TODO: replacing href on a cross-origin window should succeed?
+        Err(Error::Security)
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-location-reload
+    fn Reload(&self) -> Fallible<()> {
+        Err(Error::Security)
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-location-href
+    fn Stringifier(&self) -> DOMString {
+        DOMString::from("<XOriginWindow>")
```

Comment:
Er yes, not sure what to put here. Stringifiers aren't allowed to throw, so we have to return something. Once we have XOWs, this will never be called so we can `panic!`, but in the mean time we have to return some `DOMString`.